File: 222-CountCompleteTreeNodes/222-CountCompleteTreeNodes.java
// Last updated: 04/04/2026, 23:35:48
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void help(TreeNode root,int[] cnt) {
        if(root==null) {
            return;
        }
        cnt[0]++;
        help(root.left,cnt);
        help(root.right,cnt);
    }
    public int countNodes(TreeNode root) {
        int[] noOfNodes = new int[1];
        help(root,noOfNodes);
        return noOfNodes[0];
    }
}

File: 222-CountCompleteTreeNodes/222-CountCompleteTreeNodes.cpp
// Last updated: 04/04/2026, 23:35:41
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int leftHeight(TreeNode* root) {
        int height = 0;
        while(root) {
            height++;
            root = root->left;
        }
        return height;
    }
    int rightHeight(TreeNode* root) {
        int height = 0;
        while(root) {
            height++;
            root = root->right;
        }
        return height;
    } 
    int countNodes(TreeNode* root) {
        if(root==NULL) return 0;
        int lh = leftHeight(root->left);
        int rh = rightHeight(root->right);
        if(lh==rh) return pow(2,lh+1)-1;
        else return 1 + countNodes(root->left) + countNodes(root->right);
    }
};